File: backend/app/schemas/carryover.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field


class CarryoverSettingCreate(BaseModel):
    category_id: uuid.UUID
    carryover_type: str = Field(pattern=r"^(expire|next_month|savings|transfer|deposit)$")
    carryover_limit: Decimal | None = Field(default=None, ge=0)
    source_asset_id: uuid.UUID | None = None
    target_asset_id: uuid.UUID | None = None
    target_savings_name: str | None = Field(default=None, max_length=100)
    target_annual_rate: Decimal | None = Field(default=None, ge=0, le=100)


class CarryoverSettingUpdate(BaseModel):
    carryover_type: str | None = Field(default=None, pattern=r"^(expire|next_month|savings|transfer|deposit)$")
    carryover_limit: Decimal | None = Field(default=None, ge=0)
    source_asset_id: uuid.UUID | None = None
    target_asset_id: uuid.UUID | None = None
    target_savings_name: str | None = Field(default=None, max_length=100)
    target_annual_rate: Decimal | None = Field(default=None, ge=0, le=100)


class CarryoverSettingResponse(BaseModel):
    id: uuid.UUID
    category_id: uuid.UUID
    category_name: str
    carryover_type: str
    carryover_limit: float | None
    source_asset_id: uuid.UUID | None
    source_asset_name: str | None
    target_asset_id: uuid.UUID | None
    target_savings_name: str | None
    target_annual_rate: float | None
    created_at: datetime
    updated_at: datetime


class CarryoverLogResponse(BaseModel):
    id: uuid.UUID
    category_id: uuid.UUID
    category_name: str
    budget_period_start: date
    budget_period_end: date
    carryover_type: str
    amount: float
    target_description: str | None
    executed_at: datetime
    created_at: datetime


class CarryoverExecuteRequest(BaseModel):
    period_start: date
    period_end: date


class CarryoverPreviewResponse(BaseModel):
    category_id: uuid.UUID
    category_name: str
    carryover_type: str
    budget: float
    spent: float
    remaining: float
    carryover_amount: float
    target_description: str | None

File: frontend/src/shared/types/carryover.ts
export type CarryoverType = 'expire' | 'next_month' | 'savings' | 'transfer' | 'deposit';

export interface CarryoverSettingCreate {
  category_id: string;
  carryover_type: CarryoverType;
  carryover_limit?: number | null;
  source_asset_id?: string | null;
  target_asset_id?: string | null;
  target_savings_name?: string | null;
  target_annual_rate?: number | null;
}

export interface CarryoverSettingResponse {
  id: string;
  category_id: string;
  category_name: string;
  carryover_type: CarryoverType;
  carryover_limit: number | null;
  source_asset_id: string | null;
  source_asset_name: string | null;
  target_asset_id: string | null;
  target_savings_name: string | null;
  target_annual_rate: number | null;
  created_at: string;
  updated_at: string;
}

export interface CarryoverPreview {
  category_id: string;
  category_name: string;
  remaining_budget: number;
  carryover_type: CarryoverType;
  carryover_limit: number | null;
  target_asset_id: string | null;
  target_savings_name: string | null;
  target_annual_rate: number | null;
  source_asset_id: string | null;
  estimated_transfer_amount: number | null;
}

export interface CarryoverLog {
  id: string;
  period_start: string;
  period_end: string;
  category_id: string;
  category_name: string;
  carryover_type: CarryoverType;
  remaining_budget: number;
  transferred_amount: number | null;
  target_asset_id: string | null;
  target_savings_name: string | null;
  status: string;
  executed_at: string;
  created_at: string;
}

// Backwards-compatible aliases (legacy names)
export type CarryoverSetting = CarryoverSettingResponse;
export type CarryoverSettingRequest = CarryoverSettingCreate;
export const CARRYOVER_TYPE_LABELS: Record<CarryoverType, string> = {
  expire: '소멸',
  next_month: '다음달 이월',
  savings: '적금 저축',
  transfer: '계좌 이체',
  deposit: '예금 저축',
};
